internal/pubsub/subscriber.go

```diff
@@ -2,367 +2,232 @@ package pubsub
+	if err != nil {
+		s.logger.Errorw("failed to process data in msg", zap.Error(err))
 
-	// Ensure stream exists before we attempt to listen on it
-	if err := c.ensureStream(); err != nil {
 		return err
 	}
 
-	// Set subscription options. We specifically want manual acks in case something goes wrong
-	// persisting the event
-	subOpts := []nats.SubOpt{
-		nats.BindStream(c.stream),
-		nats.Durable(c.consumer),
-		nats.ManualAck(),
-		nats.AckExplicit(),
-	}
-
-	// For each resource type, we create a single subscription. This means that a single worker
-	// process exists for each resource type, which may be a limiting factor if some resource
-	// subjects are chattier than others
-	for _, name := range c.resourceTypeNames {
-		subject := fmt.Sprintf("%s.%s.>", c.prefix, name)
-		queueName := fmt.Sprintf("%s-%s", c.consumer, name)
+	ctx, span := tracer.Start(context.Background(), "pubsub.receive", trace.WithAttributes(attribute.String("pubsub.subject", changeMsg.SubjectID.String())))
 
-		subscription, err := c.js.QueueSubscribe(subject, queueName, c.receiveMsg, subOpts...)
-		if err != nil {
-			return err
-		}
+	defer span.End()
 
-		c.subscriptions = append(c.subscriptions, subscription)
-	}
+	resource, err := s.qe.NewResourceFromID(changeMsg.SubjectID)
+	if err != nil {
+		s.logger.Warnw("invalid subject id", "error", err.Error())
 
-	return nil
-}
+		msg.Ack()
 
-// Stop drains all subscriptions for the client.
-func (c *Client) Stop() error {
-	for _, sub := range c.subscriptions {
-		err := sub.Drain()
-		if err != nil {
-			return err
-		}
+		return err
 	}
 
-	return nil
-}
-
-func (c *Client) ackWithError(msg *nats.Msg, err error) {
-	c.logger.Warnw("invalid message - will not reprocess", "error", err.Error())
+	s.logger.Infow("received message", "resource_type", resource.Type, "resource_id", resource.ID.String(), "event", changeMsg.EventType)
 
-	err = msg.Ack()
-	if err != nil {
-		c.logger.Errorw("error acking message", "error", err.Error())
+	switch events.ChangeType(changeMsg.EventType) {
+	case events.CreateChangeType:
+		err = s.handleCreateEvent(ctx, msg, changeMsg)
+	case events.UpdateChangeType:
+		err = s.handleUpdateEvent(ctx, msg, changeMsg)
+	case events.DeleteChangeType:
+		err = s.handleDeleteEvent(ctx, msg, changeMsg)
+	default:
+		s.logger.Debugw("ignoring msg, not a create, update or delete event", zap.String("event-type", changeMsg.EventType))
 	}
-}
 
-func (c *Client) newResourceFromString(idStr string) (types.Resource, error) {
-	id, err := gidx.Parse(idStr)
 	if err != nil {
-		return types.Resource{}, err
+		return err
 	}
 
-	return c.qe.NewResourceFromID(id)
+	return nil
 }
 
-func (c *Client) createRelationships(ctx context.Context, msg *nats.Msg, resource types.Resource, fields map[string]string) error {
+func (s *Subscriber) createRelationships(ctx context.Context, msg *message.Message, resource types.Resource, additionalSubjectIDs []gidx.PrefixedID) error {
```

Comment:
So `additionalSbjectIDs` isn't the abstraction we want here - that's more of a generic "things that are related to this thing" list. I'm pretty sure what we actually want is a `map[string]gidx.PrefixedID` or something similar, where that map is built from `events.ChangeMessage.SubjectFields` or `events.ChangeMessage.Changes` as follows:

* For the given resource, look up its type in the engine schema
* For each field in the map, check to see if a relation exists with that name
* If it does, look up the target types for the relation
* If it does, ensure the prefix in the ID matches at least one prefix in the target types

As a concrete example, say we have a payload that look like so:
```json
{
    "subjectID": "loadbal-abc123",
    "subjectFields": {
        "name": "my-cool-loadbalancer",
        "owner": "tnntten-abc123"
    }
}
```

Then, assume a policy exists (in YAML) like so:
```yaml
resourceTypes:
  - name: tenant
    idPrefix: tnntten
  - name: loadbalancer
    idPrefix: loadbal
    relationships:
      - relation: owner
        targetTypeNames:
          - tenant
```

When we process the event, we should load the resource type definition from the schema. From there, we can see `owner` is a relation on the authorization policy for a `loadbalancer` resource, but `name` is not. Thus, we should persist anything called `owner` as a relationship in SpiceDB, and check that its ID prefix is `tnntten` before we do.

Reply:
This makes sense. Right now the services aren't using the `SubjectFields` attribute and is something I think we need to discuss further. I believe once we've figured that out, this will be an easy transition. For now I've published a few changes to handle the `AdditionalSubjectIDs` better.